llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -2014,6 +2011,106 @@ MCCASBuilder::createDebugStringRefs() {
+    return SF.getContents();                                                   \
+  }
+#define MCFRAGMENT_ENCODED_FRAGMENT_ONLY
+#include "llvm/MCCAS/MCCASObjectV1.def"
+  case MCFragment::FT_CVInlineLines: {
+    const MCCVInlineLineTableFragment &SF =
+        cast<MCCVInlineLineTableFragment>(Fragment);
+    return SF.getContents();
+  }
+  case MCFragment::FT_LEB: {
+    const MCLEBFragment &SF = cast<MCLEBFragment>(Fragment);
+    return SF.getContents();
+  }
+  default:
+    return ArrayRef<char>();
+  }
+}
+
+static unsigned getRelocationSize(const MachO::any_relocation_info &RE,
+                                  bool IsLittleEndian) {
+  if (IsLittleEndian)
+    return 1 << ((RE.r_word1 >> 25) & 3);
+  return 1 << ((RE.r_word1 >> 5) & 3);
+}
+
+static uint32_t getRelocationOffset(const MachO::any_relocation_info &RE) {
+  return RE.r_word0;
+}
+
+/// This helper function is used to partition a Fragment into 2 parts, the \p
+/// FinalFragmentContents will contain the contents of the Fragment that will
+/// deduplicate and will be stored as part of a *FragmentRef or a
+/// MergedFragmentRef, the \p Addends are the location of all the relocation
+/// addends that will not deduplicate in the CAS and must be stored separately,
+/// there is one AddendRef per section in the CAS. The \p RelocationBufferIndex
+/// is the index into the \p RelocationBuffer and is used to access the current
+/// relocation that has to be partitioned, we need to do this because sometimes
+/// we only have access to all the relocations for a section and a relocation
+/// may, in that case, not belong to the current fragment.
+
+/// FragmentIndex: It denotes the index into the FragmentContents that is used
+/// to copy the data that deduplicates in the \p FinalFragmentContents.
+
+/// PrevOffset: A relocation can sometimes be divided into multiple parts, all
+/// of them share the same offset, but describe only one Addend, if we already
+/// copied the addend out of the FragmentContents at a particular offset, we
+/// should skip all relocations that matches the same offset.
+
+/// RelocOffsetInFragment: This is used to denote the offset of the relocation
+/// in the current Fragment. Relocation offsets are always from the start of the
+/// section, we need to normalize this value to start at the start of the
+/// Fragment instead.
+static void
+partitionFragment(const MCAsmLayout &Layout, SmallVector<char, 0> &Addends,
+                  SmallVector<char, 0> &FinalFragmentContents,
+                  ArrayRef<MachO::any_relocation_info> RelocationBuffer,
+                  const MCFragment &Fragment, uint64_t &RelocationBufferIndex,
+                  bool IsLittleEndian) {
+  auto FragmentContents = getFragmentContents(Fragment);
+  // Index in current Fragment.
+  uint64_t FragmentIndex = 0;
+  int64_t PrevOffset = -1;
+  // Relocations are stored in the Section.
+  for (; RelocationBufferIndex < RelocationBuffer.size();
```

Comment:
Why pass in an ArrayRef + an Index instead of passing an ArrayRef where the elements before RelocationBufferIndex have been sliced off?
https://llvm.org/doxygen/classllvm_1_1ArrayRef.html#a5c498de3bb758473707e9198311eb15f

Reply:
I will still need to pass some index or number to tell me how many relocations have been read. The relocations can be section relocs, which means they belong to any fragment in the section, and therefore I need to have some way of knowing how many relocs were used up by the current fragment, the difference would be is that instead of passing an index, I would be passing a `NumOfRelocsProcessed` and then chop of the ArrayRef by that many, is that better?